### PR TITLE
Implement basic.reject frame

### LINKED ISSUE #76: Support for basic.reject frame
### Description

The basic.reject frame can be used by a client to reject an incoming message. This method allows a client to reject a message. It can be used to interrupt and cancel large incoming messages, or return untreatable messages to their original queue.

#### Guidelines for implementers: 
- The server SHOULD be capable of accepting and process the Reject method while sending message content with a Deliver or Get­Ok method. I.e. the server should read and process incoming methods while sending output frames. To cancel a partially­send content, the server sends a content body frame of size 1 (i.e. with no data except the frame­end octet). 
- The server SHOULD interpret this method as meaning that the client is unable to process the message at this time. 
- The client MUST NOT use this method as a means of selecting messages to process.

### PR DESCRIPTION
**Pull request by a5anka**
Handle message rejection
Fixes #76

Implement basic.reject frame handling logic.

Implement basic.frame handle logic. If the requeue bit is set the message is enqueued again to the queue.

As an application, if the message cannot be processed it should be able t o put the message back in to the queue.

- Implement basic.reject frame for rejecting and requeueing messages

N/A . AMQP documentation already has this.

N/A

N/A

N/A

## Automation tests
 - Unit tests 73%
   > Code coverage information
 - Integration tests - in progress

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java [hunk 1]
```diff
@@ -124,14 +124,20 @@ Message dequeue() {
...
     void acknowledge(long messageId) {
         // TODO handle nacks
-        pendingMessages.remove(messageId);
+        Message message = pendingMessages.remove(messageId);
+        message.release();
+    }
+
+    public void requeue(long messageId) {
+        Message message = pendingMessages.remove(messageId);
+        queue.enqueue(message);
```
Comment (Asitha) [inline modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java]: So we are putting the message into end of the queue? So it get delivered much later. Specification doesn't say anything on this right?

File: modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java [hunk 2]
```diff
@@ -124,14 +124,20 @@ Message dequeue() {
...
 
     void acknowledge(long messageId) {
         // TODO handle nacks
-        pendingMessages.remove(messageId);
+        Message message = pendingMessages.remove(messageId);
+        message.release();
+    }
+
+    public void requeue(long messageId) {
+        Message message = pendingMessages.remove(messageId);
```
Comment (Asitha) [inline modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java]: What about the redelivered flag on the message?

Comment (a5anka) [reply modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java]: Fix done in 3fa16258b5751df6154b9a248ce577bbc09766ad

File: modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java (hunk 1 above)
Comment (a5anka) [reply modules/broker-core/src/main/java/org/wso2/broker/core/QueueHandler.java]: Noted. Created https://github.com/wso2/message-broker/issues/93 to track this

**Pull request merged**